### PR TITLE
feat(ingestion): adds the optional instance field if given when producing the container entities

### PR DESCRIPTION
**Pull request by sgomezvillamor**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by sgomezvillamor** [comment]
@treff7es you were working on some stuff related to this recently, so sharing with you

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 1]
```diff
@@ -275,15 +275,16 @@ def _extract_record(self, topic: str) -> Iterable[MetadataWorkUnit]:  # noqa: C9
...
 
         # 4. Attach dataPlatformInstance aspect.
-        if self.source_config.platform_instance:
-            dataset_snapshot.aspects.append(
-                DataPlatformInstanceClass(
-                    platform=platform_urn,
-                    instance=make_dataplatform_instance_urn(
-                        self.platform, self.source_config.platform_instance
-                    ),
+        dataset_snapshot.aspects.append(
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/kafka.py:278]: This makes sense!

File: metadata-ingestion/src/datahub/emitter/mcp_builder.py [hunk 2]
```diff
@@ -167,6 +171,9 @@ def gen_containers(
         aspectName="dataPlatformInstance",
         aspect=DataPlatformInstance(
             platform=f"{make_data_platform_urn(container_key.platform)}",
+            instance=f"{make_dataplatform_instance_urn(container_key.platform, container_key.instance)}"
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/emitter/mcp_builder.py:174]: I think we should instead add a new argument to the function itself and use that - I recall there is some inconsistency among sources in terms of how the "instance" field is populated :( 

cc @treff7es To take a look!

Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/emitter/mcp_builder.py:174]: I know in some cases _environment_ values (dev, ei, prod, ...) are used for the _platform instance_. This was decided to keep backward compatibility with an issue that was introduced previously in some connector and so to avoid migrating URNs. Is that what you mean?

Anyway, not clear for me if this is a comment blocking this PR or just an observation for a future improvement. If there is something specific I could do in this PR, just let me know please.

**Comment by sgomezvillamor** [comment]
Hi, any update? /cc: @treff7es @jjoyce0510

**Comment by treff7es** [comment]
> Hi, any update? /cc: @treff7es @jjoyce0510

I just checked and basically what we do is to generate PlatformInstance even if the platform is not set in the config or not supported by the source.
I need to ask around if this generalization makes sense for all the sources because afaik there are sources where the PlatformInstance does not make sense.

**Comment by sgomezvillamor** [comment]
Unless I'm missing something 😅 ...

> I just checked and basically what we do is to generate PlatformInstance even if the platform is not set in the config or not supported by the source.

If the platform is not set in the config or not supported by the source, the PlatformInstance aspect for the container upsert event was already generated but only with the `platform` detail. This PR just adds the `instance` detail for those cases where the platform instance is supported by the source and set in the config.

> I need to ask around if this generalization makes sense for all the sources because afaik there are sources where the PlatformInstance does not make sense

In those cases, the PlatformInstance aspect for the container upsert will include only the `platform` not the instance (because in `container_key.instance` should be `null`


Basically, what I want to solve in this PR is the following misalignment:
* datasets exist in the context of a platform instance
* at the same time, those datasets are defined in the context of containers
* however, those containers do not hold any reference to the parent platform instance (well, actually they do, in the encoded container id 😅 )
Instead the PR tries to make platform instance info easily accessible for container entities.

**Comment by sgomezvillamor** [comment]
Hi, is there any update on this? /cc: @treff7es @jjoyce0510

I could split the PR in two, if that helps to move this forward somehow:
* on the one side, align the behaviour of the _kafka_ connector with the _kafka connect_ one, as pointed out here https://github.com/datahub-project/datahub/pull/4665#discussion_r849822074
* on the other, adding the platform instance detail for containers to fix the misalignment pointed out here https://github.com/datahub-project/datahub/pull/4665#issuecomment-1105688384

Fixing conflicts is just postponed 😅 until we have confirmation these updates make sense to you too.

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 3]
```diff
@@ -298,15 +298,16 @@ def _extract_record(self, topic: str) -> Iterable[MetadataWorkUnit]:
...
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/kafka.py:301]: We need to rethink what is the invariant we want when we find a data-platform-instance aspect attached to an entity. Is it that the recipe author provided a specific platform_instance to use? or that the ingestion code auto-generated an instance based on the ENV. 

The dataplatform instance migration script depends on finding no data platform instance aspect or finding a data platform instance without the instance specified, to figure out which entities to migrate... so it might get trickier if we have lots of entities sitting around with "default instances" being populated. 

We shouldn't accept this PR until we have figured out what the assertion is.

**Review by shirshanka — CHANGES_REQUESTED** [review]
No changes requested, just a comment indicating that we need to get back to @sgomezvillamor with what guarantees we want to have in the metadata graph w.r.t data-platform-instance values.

**Comment by hsheth2** [comment]
@sgomezvillamor I'm working on https://github.com/datahub-project/datahub/pull/6083, which should accomplish the end goal as this PR while maintaining backwards compat a bit better.

**Comment by sgomezvillamor** [comment]
Superseded by https://github.com/datahub-project/datahub/pull/6083
